### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by arkhipowasilii**
Что здесь надо писать?

### CODE REVIEW

**Comment by cyphersnake** [comment]
В описание PR писать что с собой несут изменения в данном PR. Желательно на английском.

File: bot.py [hunk 1]
```diff
@@ -81,46 +74,71 @@ def open_particular_products(self, update: Update, context,
...
+            self.edit_message_reply_markup(update, context, kb)
+
+    def get_basket(self, update: Update, context, *args):
+        max_offset, product_data = self._catalog.get_basket(update._effective_user.id, 0, 1)
+        kb = KB(self._serializer)
+        kb.button("delete", self.delete_products_from_cart, (0,)).\
+            pager(callback=self.get_basket2, in_page=1, current_offset=0, max_offset=max_offset)
+        self.send_message_photo(update=update, context=context, photo=product_data[2], caption=product_data[1], kb=kb)
+
+    def get_basket2(self, update, context, offset: int, ):
```
Comment (cyphersnake) [inline bot.py]: `get_basket2`? Remove it or change name

File: catalog/db.py [hunk 2]
```diff
@@ -96,6 +96,35 @@ def update_column(self, column: str, value: int, id: int):
         self._connect.cursor().execute(sql)
         self._connect.commit()
 
+    def insert_into_basket(self, user_id: int, good_id: int, count: int):
+        sql1 = (f"SELECT EXISTS(SELECT 1 FROM basket "
+                f"WHERE user_id = {user_id} AND good_id = {good_id})")
+        if self._connect.cursor().execute(sql1).fetchone()[0]:
```
Comment (cyphersnake) [inline catalog/db.py]: exception, if data does't exist

File: catalog/db.py [hunk 3]
```diff
@@ -96,6 +96,35 @@ def update_column(self, column: str, value: int, id: int):
         self._connect.cursor().execute(sql)
         self._connect.commit()
 
+    def insert_into_basket(self, user_id: int, good_id: int, count: int):
+        sql1 = (f"SELECT EXISTS(SELECT 1 FROM basket "
```
Comment (cyphersnake) [inline catalog/db.py]: If you have more than one sql request - give them normal names

File: catalog/db.py [hunk 4]
```diff
@@ -96,6 +96,35 @@ def update_column(self, column: str, value: int, id: int):
...
+        self._connect.commit()
+
+    def get_basket(self, user_id: int, offset: int, limit: int) -> Union[int, Iterable[Tuple[int, str, str, int]]]:
+        sql1 = f"SELECT gd.id, gd.name, gd.photo_id, ba.count FROM basket as ba "\
+            f"LEFT JOIN goods as gd ON ba.good_id = gd.id " \
+            f"WHERE ba.user_id = {user_id} " \
+            f"ORDER BY gd.name " \
+            f"LIMIT {limit} OFFSET {offset}"
+
+        sql2 = f"SELECT count(*) FROM basket " \
```
Comment (cyphersnake) [inline catalog/db.py]: You can do it in one request

File: service/keyboard_builder.py [hunk 5]
```diff
@@ -60,3 +60,22 @@ def get(self) -> Markup:
...
+        self._serializer = serializer
+        self._buttons = []
+        self._buttons.append([])
+
+    def button(self, text: str, callback: Callable, data: Iterable[Any] = None) -> 'MenuBuilder':
+        data = self._serializer.serialize(callback, data or [])
+        self._buttons[-1].append(MenuButton(text=text, callback_data=data))
+        return self
+
+    def get(self) -> MenuMarkup:
```
Comment (cyphersnake) [inline service/keyboard_builder.py:83]: Add `line` function

File: bot.py [hunk 6]
```diff
@@ -32,15 +32,8 @@ def _connect(self):
...
-        photo = self.open_main_photo()
         kb = KB(self._serializer).button("Open", self.open_categories)
-        msg = self.send_message_photo(update,
-                                      context,
-                                      "I'm catalog! Insert your request or open categories!",
-                                      kb,
-                                      photo)
-        main_image_path = msg.photo[-1].file_id
-        photo.close()
+        self.send_message(update, context, "I'm catalog! Insert your request or open categories!", kb)
```
Comment (cyphersnake) [inline bot.py]: You can send `MenuKeyboard` here

**Review by cyphersnake — COMMENTED** [review]
Fix it before leeson

File: service/keyboard_builder.py (hunk 5 above)
Comment (arkhipowasilii) [reply service/keyboard_builder.py:83]: After self?

File: bot.py (hunk 1 above)
Comment (arkhipowasilii) [reply bot.py]: done

File: catalog/db.py (hunk 2 above)
Comment (arkhipowasilii) [reply catalog/db.py]: exception already exists

File: catalog/db.py (hunk 4 above)
Comment (arkhipowasilii) [reply catalog/db.py]: How to do it?

File: catalog/db.py (hunk 2 above)
Comment (cyphersnake) [reply catalog/db.py]: No, it does't work well

File: catalog/db.py (hunk 4 above)
Comment (arkhipowasilii) [reply catalog/db.py]: count(*) will count rows from table goods and basket

File: service/keyboard_builder.py (hunk 5 above)
Comment (cyphersnake) [reply service/keyboard_builder.py:83]: Like in the another builder

File: catalog/db.py (hunk 3 above)
Comment (arkhipowasilii) [reply catalog/db.py]: done

File: bot.py (hunk 1 above)
Comment (cyphersnake) [reply bot.py]: Waiting for push in this branch

File: catalog/db.py (hunk 2 above)
Comment (cyphersnake) [reply catalog/db.py]: And why `()` is around a string?

File: catalog/db.py (hunk 4 above)
Comment (cyphersnake) [reply catalog/db.py]: Ok, with limit and offset it really cannot be done in one request

File: catalog/db.py [hunk 7]
```diff
@@ -96,6 +96,33 @@ def update_column(self, column: str, value: int, id: int):
...
+        return self._connect.cursor().execute(sql2).fetchone()[0], self._connect.cursor().execute(sql1).fetchall()
+
+    def delete_product_from_basket(self, user_id: int, good_id: int):
+        sql = f"DELETE FROM basket WHERE user_id = {user_id} AND good_id = {good_id}"
+        self._connect.cursor().execute(sql)
+        self._connect.commit()
+
 
 if __name__ == '__main__':
     db = DatabaseHandler(Path(r'/Users/arkhipowasilii/PycharmProjects/catalog-bot2/catalog.db'))
```
Comment (cyphersnake) [inline catalog/db.py]: Remove it and don't commit it in repository anymore

File: catalog/db.py [hunk 8]
```diff
@@ -96,6 +96,33 @@ def update_column(self, column: str, value: int, id: int):
         self._connect.cursor().execute(sql)
         self._connect.commit()
 
+    def insert_into_basket(self, user_id: int, good_id: int, count: int):
+        sql_update = f"UPDATE basket SET count = count + {count} " \
+            f"WHERE user_id = {user_id} AND good_id = {good_id}"
+        sql_insert = f"INSERT INTO basket (user_id, good_id, count) " \
+            f"SELECT {user_id}, {good_id}, 1 " \
+            f"WHERE (SELECT Changes() = 0)"
```
Comment (cyphersnake) [inline catalog/db.py:113]: Comment this magic

File: catalog/catalog.py [hunk 9]
```diff
@@ -71,3 +71,13 @@ def update_photo_id(self, photo_id: int, id: int):
 
     def get_description(self, id: int) -> str:
         return self._db.get_description(id)
+
+    def insert_into_basket(self, user_id: int, good_id: int, count: int):
+        self._db.insert_into_basket(user_id, good_id, count)
+
+    def get_basket(self, user_id: int, offset: int, limit: int) -> Iterable[Product]:
```
Comment (cyphersnake) [inline catalog/catalog.py]: Fix return type

**Comment by cyphersnake** [comment]
Rename all `basket` to `card` (in db too)

**Pull request merged**